### DISCUSSION TITLE
Support for .bib files

### DISCUSSION THREAD

**Discussion by Idobenhamo** [Feel-ix-343/markdown-oxide#12] [Ideas]
Hi, 
Adding support for citations can be awesome!

**Comment by Feel-ix-343** [comment]
Thanks for the comment!

Could you explain your workflow for markdown and .bib files? I am quite interested!

**Reply by Idobenhamo** [reply]
I believe a basic citation system that utilizes `@` and automatically links citations to the corresponding entry in the bibliography file would be fantastic!

By the way, I was wondering if it's possible to link non-Markdown files, like PDFs and such, in my "vault"? If I have other file types like typst files, will they coexist peacefully with the LSP and MD files? Perhaps setting up a Discord server or similar could be beneficial...

Thanks a lot for your incredible work and for sharing it with us. I hope I can contribute in the future when I have some free time!

**Reply by Feel-ix-343** [reply]
I like this idea a lot; it would be especially helpful for me when writing papers in school. 

I would want to keep this syntax compatible with Obsidian, so I think we should use footnotes to somehow integrate with .bib. The following workflow is appealing to me
1. Type [ and get .bib completions
2. Fuzzy match and select
3. When selected, a footnote will be completed: e.g [^bibid] and a new footnote item will be appended to the file. This footnote item will have an mla/apa/... (select in config) citation generated from the .bib code. The footnote could also have something more useful that I am not thinking of; do let me know what you think!

Also I just remembered this library: https://github.com/typst/biblatex. If you're up for it, see if you can get something working based on that! I would really appreciate some help on this repo. 

As for the other files, right now the LS does not pay any attention to files other than markdown files. However, you bring up a good idea about completions for images and pdfs and such. I don't think it would be too hard to implement as well. So no a vault with pngs won't break anything but you won't get completions/references/thelike for them.

And for the discord, I personally really like the organization and centralization of github-discussions. Don't hesitate to ask as many questions or give as many suggestions as you want to -- we are trying to create the best PKM language server after all.

**Reply by Feel-ix-343** [reply]
Hey @Idobenhamo how do you feel about this? Bumping in case you did not see it.

**Reply by Idobenhamo** [reply]
Apologies for the delayed response; I'm currently swamped with work and have no spare time at the moment.

Regarding your suggestion, using `[` to suggest all the bib files might become messy with longer `bib` files. Perhaps using `[@` could be a better alternative?

I also think the footnote idea is great! Maybe even making the DOI address a clickable link to the article could enhance the feature.

I appreciate your hard work on this project! Once I have some free time, I'm excited to learn some Rust and join you. By the way, if you have any recommendations on the fastest way for me to get up to speed with Rust to contribute to this project, I'd love to hear them.

**Reply by Feel-ix-343** [reply]
The [rust book](https://doc.rust-lang.org/book/title-page.html) is 100% the best IMO. I am excited for you to eventually join!

I do have to say some of the code is a little messy as I was trying to get this rolled out as fast as possible. However, I recently refactored the completions API [here](https://github.com/Feel-ix-343/markdown-oxide/pull/33) and the `vault` API is pretty good IMO too. I'd love to talk through the design of it any time you want to; you could book me by my cal.com link in my bio. To encourage contributions I will also make a few wiki entries explaining things.

**Reply by Idobenhamo** [reply]
What is next feature planned?
Wanted to say thanks again for your awesome work!

**Reply by Feel-ix-343** [reply]
Hey right now I am doing a major refactor, then fix a few bugs and publish the project to hacker-news/reddit. After that some of the more specialized features like this and better, llm powered completions will be made.

Thanks for checking in!

**Reply by Idobenhamo** [reply]
BTW, How the wiki links work with files inside folders? I think there is something weird with that....
There is no Completions for relative file path and the linking works also without the relative path using only the filename.
Thanks you again for your awesome work!

**Reply by Feel-ix-343** [reply]
Hey so I talked about this a lot in #2, but basically we decided to use filenames, and not relative paths as the default. You can still link to files in folders but just by using the file names. The main reason this decisions was made was that it was more robust for when files are moved around - you dont need to change all of your links to them.

This being said, relative paths still are resolved, but you just dont get completions for them.

And soon I will add an option in settings for relative path completions.

**Reply by Idobenhamo** [reply]
Could it be acceptable to include `@` in a filename without causing potential issues in the future? For instance, using `@example.md` and linking it as `[[@example]]` for literature notes seems like a convenient workflow. 
Thanks for your continued assistance!

**Reply by Feel-ix-343** [reply]
Yea I dont see why not

You could also use aliases (check readme)

Note that the nvim cmp keyword pattern may need to be modified to support this. I'm not sure if it would work right now - can't check either.

If it's not supported, Could you open an issue?
